### ISSUE TITLE
MSAL says broker is not enabled for device

### ISSUE THREAD

**Issue by geohei** [AzureAD/microsoft-authentication-library-for-objc#1523]
I'm using sample package below to make myself familiar with MSAL 1.2.2 (MSALiOS).
https://github.com/Azure-Samples/ms-identity-mobile-apple-swift-objc
Microsoft Authenticator is installed on the device and correctly used by sample code above.
All works fine.

But ... if I implement the code in my app, I get an error while getting device information:

```
msalApplicationContext?.getDeviceInformation(with: nil, completionBlock: { (deviceInformation, error) in
    guard let deviceInfo = deviceInformation else { return }
    msalCurrentDeviceMode = deviceInfo.deviceMode
}
```

`Optional(Error Domain=MSALErrorDomain Code=-50000 "(null)" UserInfo={MSALErrorDescriptionKey=Broker is not enabled for this operation. Please make sure you have enabled broker support for your application, MSALInternalErrorCodeKey=-42714})`

The prerequisites are set:
1. Keychain Sharing
`com.microsoft.adalcache`
2. Info.plist
```
<key>LSApplicationQueriesSchemes</key>
<array>
    <string>msauthv2</string>
    <string>msauthv3</string>
</array>
```
3. AppDelegate callback function is implemented.

`MSALError.h` says:
```
/*
Broker is either not found on device or not available for this configuration.
*/
MSALInternalBrokerNotAvailable                      = -42714

```
Am I missing something or did the docs miss to mention something?
In other words ... when does `getDeviceInformation()` return an error?

**Comment by kaisong1990** [comment]
@geohei I would suggest going to ```MSIDRequestParameters+Broker.m```, and put breakpoint in the class to check which validation failed

**Comment by geohei** [comment]
@kaisong1990 Thanks for your reply. Meanwhile I managed to make the broker (Microsoft Authenticator) pop up. I believe my mistake was that I left `msauth.com.microsoft.identitysample.MSALiOS://auth` (from the https://github.com/Azure-Samples/ms-identity-mobile-apple-swift-objc sample package) in the `Info.plist` (URL types) as well as `kRedirectUri`. When I changed both settings to the `msauth.MY_BUNDLE_ID://auth` of my iOS app, it worked. Is my (!) `MY_BUNDLE_ID` supposed to be setup on the Azure AD server configuration to get full authentication services (e.g. broker), or can any (iOS) app benefit from the full range of services without that Azure AD knows which client (iOS app) authenticates?


Can you confirm that `MY_BUNDLE_ID` needs to be put in `Info.plist` and `kRedirectUri` in order to make a broker work?

**Comment by kaisong1990** [comment]
@geohei You might find this link useful as a quick start:
https://docs.microsoft.com/en-us/azure/active-directory/develop/mobile-app-quickstart?pivots=devlang-ios

```MY_BUNDLE_ID``` was added into info.plist by Xcode when creating the app. And yes, the redirectUri needs to be added manually into infp.plist, and it needs to be the same as the one you registered in the portal in order to make the auth flow work

**Comment by geohei** [comment]
@kaisong1990 I followed your quick start link, created an Azure test account and tested quite some configurations. The problem with this project is that (1.) I do not have access to the Azure AD I have to write iOS code for, (2.) nor do I have an test user account to test run my settings/configuration. All I have is the concerned Azure's application URL answer from a user's Firefox web inspector return.

URL answer:
`https://login.microsoftonline.com/11111111-1111-1111-1111-111111111111/login`

```
{
	"GET": {
		"scheme": "https",
		"host": "login.microsoftonline.com",
		"filename": "/11111111-1111-1111-1111-111111111111/oauth2/authorize",
		"query": {
			"response_type": "code",
			"client_id": "22222222-2222-2222-2222-222222222222",
			"scope": "openid",
			"nonce": "33333333-3333-3333-3333-333333333333",
			"redirect_uri": "https://service.company.net/service/",
			"state": "AppProxyState:{\"InvalidTokenRetry\":null,\"IsMsofba\":false,\"OriginalRawUrl\":\"https:\\/\\/serice.company.net\\/service\",\"RequestProfileId\":null,\"SessionId\":\"55555555-5555-5555-5555-555555555555\"}#EndOfStateParam#",
			"client-request-id": "55555555-5555-5555-5555-555555555555"
		},
		"remote": {
			"Address": "1.2.3.4:443"
		}
	}
}

```
What I did in Xcode with this:

```
kClientID = "22222222-2222-2222-2222-222222222222"
kRedirectURI = "https://service.company.net/service/"
kGraphEndpoint = "https://graph.microsoft.com/"
kAuthority = "https://login.microsoftonline.com/11111111-1111-1111-1111-111111111111"
kScopes = ["user.read"]

```
The Info.plist holds `msauth.[MY_BUNDLE_ID]` as URL types.

Is the URL answer above correctly translated into the Xcode constants?

The aim is, that my app reacts the same way as the Graph API / Authenticator when Safari is used.

**Comment by kaisong1990** [comment]
@geohei I see, unfortunately, we don't support both Safari redirect and broker flow redirect at the same time. In order to use broker flow, you will need to update the redirectUri from Azure portal (or ask your team's admin for the updates).

**Comment by kaisong1990** [comment]
Closing for no activities
